Show loading indicator for rating stars

kumar303 opened the pull request:
Fixes https://github.com/mozilla/addons-frontend/issues/6014

<details>
<summary>Large stars example</summary>

![addons-frontend-stars-big mov](https://user-images.githubusercontent.com/55398/44537992-bb8abd00-a6c5-11e8-9b02-ad167a4f8989.gif)

</details>


<details>
<summary>Small stars example</summary>

![addons-frontend-stars-small mov](https://user-images.githubusercontent.com/55398/44538020-ccd3c980-a6c5-11e8-862b-4e72887ccb8a.gif)

</details>

bobsilverberg commented on src/ui/components/UserRating/index.js:
```diff
@@ -12,6 +12,7 @@ import type { UserReviewType } from 'amo/actions/reviews';
 type Props = {|
   className?: string,
   isOwner?: boolean,
+  loading?: boolean,
```
Wouldn't `loading` just be the same as `!readOnly && review === undefined`? Or do you need this for testing purposes?

kumar303 replied:
The `RatingManager` needs to present stars that the user can select to rate the component. In this case, `rating==undefined` because no rating has been set. My intention was to let `RatingManager` control the loading state but let me double check to see I still need to do that.

bobsilverberg commented on tests/unit/ui/components/TestRating.js:
```diff
@@ -228,7 +239,7 @@ describe(__filename, () => {
 
   it('selects stars on hover', () => {
-    const root = render();
+    const root = render({ loading: false });
```
You mostly use `renderWithRating()` to end up with a component that isn't in a loading state in these tests (including in "does nothing when you hover over stars", below), but for these two tests you use `render({ loading: false })`. Is there a reason for that difference in these two tests?

kumar303 commented on src/amo/components/RatingManager/index.js:
```diff
@@ -196,10 +196,13 @@ export class RatingManagerBase extends React.Component<InternalProps, State> {
     const isLoggedIn = Boolean(userId);
 
+    invariant(addon, 'addon is required');
+    invariant(version, 'version is required');
```
I removed an if statement that was checking the existence of `addon`. It wasn't necessary because `addon` is required. However, `<Addon>` is the caller of `<RatingManager>` and that component isn't covered by Flow so I thought it would be a good idea to add some invariant protection. While I was at it, I added protection for `version` which is the only other required prop.

kumar303 commented:
Thanks. I think I addressed all comments. Can you take another look?

In the previous approach, I added the `loading` prop to allow a calling component to control the loading state in addition to an implicit loading state. This adds a dimension of complexity but I thought I needed it. 

After your comments, I took a fresh look and found out how to do everything with an implicit loading state. This simplified the logic in `Rating`. One drawback is it makes the calling code a bit less explicit but that's probably OK.

bobsilverberg commented on tests/unit/amo/components/TestRatingManager.js:
```diff
@@ -124,6 +124,38 @@ describe(__filename, () => {
+
+    const root = render({ addon, store });
+
+    expect(root.find(UserRating)).toHaveProp('review', undefined);
```
I see what you mean about the impact of your recent change. The code does seem cleaner, but it's a bit less clear exactly how it's working. For example, in this test, it's described as "renders loading stars..." but doesn't have an assertion that seems to match (i.e., the assertion doesn't say anything about loading, rather it's about the contents of the `review` prop).

Maybe the descriptions of these tests should be updated to describe what we actually expect, which is a specific value for the `review` prop?

kumar303 replied:
Yeah, I can see how changing these descriptions could be helpful. The loading state is now the responsibility of `Rating`.
